modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/GridCacheDatabaseSharedManager.java

```diff
@@ -1485,69 +1485,89 @@ private void prepareIndexRebuildFuture(int cacheId) {
 
-                    if (rebuildFut != null) {
-                        log().info("Started indexes rebuilding for cache [name=" + cacheCtx.name()
-                            + ", grpName=" + cacheCtx.group().name() + ']');
+        for (GridCacheContext cacheCtx : (Collection<GridCacheContext>)cctx.cacheContexts()) {
+            if (!cacheCtx.startTopologyVersion().equals(exchangeFut.initialVersion()))
+                continue;
 
-                        if (compoundAllIdxsRebuilt == null)
-                            compoundAllIdxsRebuilt = new GridCompoundFuture();
+            int cacheId = cacheCtx.cacheId();
+            GridFutureAdapter<Void> usrFut = idxRebuildFuts.get(cacheId);
 
-                        compoundAllIdxsRebuilt.add(rebuildFut);
+            if (isNull(rebuildIdxExecSvc))
+                rebuildIdxExecSvc = qryProc.getIndexing().rebuildIndexExecutorService(0);
 
-                        rebuildFut.listen(new CI1<IgniteInternalFuture>() {
-                            @Override public void apply(IgniteInternalFuture fut) {
-                                idxRebuildFuts.remove(cacheId, usrFut);
+            IgniteInternalFuture<?> rebuildFut = qryProc.rebuildIndexesFromHash(cacheCtx, rebuildIdxExecSvc);
 
-                                Throwable err = fut.error();
+            if (nonNull(rebuildFut)) {
+                if (log.isInfoEnabled())
+                    log.info("Started indexes rebuilding for cache [" + cacheInfo(cacheCtx) + ']');
 
-                                usrFut.onDone(err);
+                if (isNull(allCacheIdxsCompoundFut))
+                    allCacheIdxsCompoundFut = new GridCompoundFuture<>();
 
-                                CacheConfiguration ccfg = cacheCtx.config();
+                allCacheIdxsCompoundFut.add(rebuildFut);
 
-                                if (ccfg != null) {
-                                    if (err == null) {
-                                        if (log.isInfoEnabled())
-                                            log.info("Finished indexes rebuilding for cache [name=" + ccfg.getName()
-                                                + ", grpName=" + ccfg.getGroupName() + ']');
-                                    }
-                                    else {
-                                        if (!(err instanceof NodeStoppingException))
-                                            log().error("Failed to rebuild indexes for cache  [name=" + ccfg.getName()
-                                                + ", grpName=" + ccfg.getGroupName() + ']', err);
-                                    }
-                                }
-                            }
-                        });
+                rebuildFut.listen(fut -> {
+                    idxRebuildFuts.remove(cacheId, usrFut);
+
+                    Throwable err = fut.error();
+
+                    usrFut.onDone(err);
+
+                    if (isNull(err)) {
```

Comment:
These methods are indented to use in predicates.
Using them in _if_ conditions looks weird.

Reply:
I think it doesn't look like a mistake, I leave it as it is.